opt: support ordered aggregations using window functions

ridwanmsharif opened the pull request:
Fixes #23620.

This change adds functionality to support ordered aggregations.
If any aggregation is non-commutative (array_agg, concat_agg) and
it specifies an ordering of results, they are now supported.
We do this by building all the aggregations as window functions
where we specify the ordering as part of the window expression.

For example:
Consider Table `tab` with the following composition:
```
col1 | col2
+------+------+
    10 |    2
    20 |    1
(2 rows)
```

The query `SELECT array_agg(col1 ORDER BY col1), array_agg(col1
ORDER BY col2) FROM tab` will return the following row preserving
the user specified orderings:
```
  array_agg | array_agg
+-----------+-----------+
  {10,20}   | {20,10}
(1 row)
```

Release note: None

Commit:
opt: support ordered aggregations using window functions

Fixes #23620.

This change adds functionality to support ordered aggregations.
If any aggregation is non-commutative (array_agg, concat_agg) and
it specifies an ordering of results, they are now supported.
We do this by building all the aggregations as window functions
where we specify the ordering as part of the window expression.

For example:
Consider Table `tab` with the following composition:
```
col1 | col2
+------+------+
    10 |    2
    20 |    1
(2 rows)
```

The query `SELECT array_agg(col1 ORDER BY col1), array_agg(col1
ORDER BY col2) FROM tab` will return the following row preserving
the user specified orderings:
```
  array_agg | array_agg
+-----------+-----------+
  {10,20}   | {20,10}
(1 row)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38014)
<!-- Reviewable:end -->

andy-kimball commented on pkg/sql/opt/optbuilder/groupby.go:
```diff
@@ -108,6 +119,16 @@ func (a *aggregateInfo) TypeCheck(ctx *tree.SemaContext, desired *types.T) (tree
 }
 
+// IsCommutative checks whether the aggregate is commutative or ordering insensitive.
+func (a aggregateInfo) IsCommutative() bool {
```
Ideally, we could reuse `AggregateIsOrderingSensitive` here so we don't have multiple places where we list commutative aggregates.

Commit:
opt: wrap scalar group by with project for window aggregations

This change adds a projection over the scalar group by for
any aggregations build as window functions. This is so the
default values for aggregations are preserved when applying
the aggregation over an empty relation.

Release note: None

rytaft reviewed (changes requested):
Looks great so far! Haven't gotten a chance to review everything, but left some comments to start.

> Reviewed 5 of 13 files at r1, 1 of 4 files at r2, 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-:-Lgitrd3DlOQCEFCiJAp:byf25mq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @ridwanmsharif)

___
*[pkg/sql/lease.go, line 474 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgif_A_2ah79J-KRlT0:-Lgif_A_2ah79J-KRlT1:b-5oexeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/lease.go#L474)):*
> ```Go
> 		ctx, "count-leases", nil /* txn */, stmt, at.GoTime(),
> 	)
> ```

stray extra line
___
*[pkg/sql/logictest/testdata/logic_test/aggregate-opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgijRdDANXBd_cfG04S:-LgijRdE8hRjgdnAGTLb:b-lhecig) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/logictest/testdata/logic_test/aggregate-opt#L1)):*
> ```
> # LogicTest: local-opt fakedist-opt
> ```

nice tests! Maybe add a couple with concat_agg and string_agg too
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgijcuPAkyzuw3NgTr6:-LgijcuPAkyzuw3NgTr7:b-o2rmsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/logictest/testdata/logic_test/orms#L27)):*
> ```
> # Ordered aggregations are possible.
> # #12115
> # Skipped until #25148 is solved
> ```

You can take out this comment now. Should you also add #25148 to the list of issues that this PR fixes?
___
*[pkg/sql/opt/optbuilder/scope.go, line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgikswTDdeg56lmk2_C:-LgikswTDdeg56lmk2_D:bd6oyzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/scope.go#L506)):*
> ```Go
> 
> 					// Now check if the orderings are identical.
> 					if match {
> ```

[nit] I think this would be slightly more efficient:
```
if match && !agg.IsCommutative() {
  if len(a.OrderBy) != len(agg.OrderBy) {
    match = false
  } else {
    for k := range a.OrderBy {
    ...
```
___
*[pkg/sql/opt/optbuilder/scope.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgildIACz9HtGf2nguD:-LgildIACz9HtGf2nguE:bjtwftc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/scope.go#L507)):*
> ```Go
> 					// Now check if the orderings are identical.
> 					if match {
> 						for k := range a.OrderBy {
> ```

[super nit] I'd use `j` here instead of `k` since you're only one level deep, not two 
___
*[pkg/sql/opt/optbuilder/select.go, line 696 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgiloZHAmlhskfceRk3:-LgiloZHAmlhskfceRk4:b-bv9iq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8acfe8ffd0028ce1d81a9b1148f7e9ba2673bf95/pkg/sql/opt/optbuilder/select.go#L696)):*
> ```Go
> 	b.buildDistinctOnArgs(fromScope, projectionsScope, distinctOnScope)
> 	b.buildProjectSet(fromScope)
> ```

stray deleted line
___
*[pkg/sql/opt/optbuilder/window.go, line 67 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgilxN02Z_VPFo14xcQ:-LgilxN02Z_VPFo14xcR:b-8nyncm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L67)):*
> ```Go
> 
> var unboundedEndBound = &tree.WindowFrameBound{BoundType: tree.UnboundedFollowing}
> var unboundedStartBound = &tree.WindowFrameBound{BoundType: tree.UnboundedPreceding}
> ```

[super nit] put start before end
___
*[pkg/sql/opt/optbuilder/window.go, line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgio9_73iGtcmwspMzi:-Lgio9_73iGtcmwspMzj:bj9kw19) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L193)):*
> ```Go
> 	for i := range groupingCols {
> 		groupingColSet.Add(int(groupingCols[i].id))
> 	}
> ```

The calling function, `buildAggregation`, constructs this same `ColSet` later on in that function. I'd move that construction above the call to `buildAggregationAsWindow` so you can just pass in the `ColSet`.
___
*[pkg/sql/opt/optbuilder/window.go, line 213 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgiq0zwCb_tzRFoPPGD:-Lgiq0zwCb_tzRFoPPGE:b-zdq5mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L213)):*
> ```Go
> 			ord := make(opt.Ordering, 0, len(agg.OrderBy))
> 			for j, t := range agg.OrderBy {
> 				te := argScope.resolveAndRequireType(t.Expr, types.Any)
> ```

If you're passing `types.Any`, you should probably just use `resolveType`
___
*[pkg/sql/opt/optbuilder/window.go, line 227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgiqSNaD8eX_tcW-Ba9:-LgiqSNaD8eX_tcW-BaA:bv6l13u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L227)):*
> ```Go
> 						)
> 					}
> 					ord = append(ord, opt.MakeOrderingColumn(col.id, t.Direction == tree.Descending))
> ```

This stuff can probably be pulled out into a helper function that's called both here and above in `buildWindow`
___
*[pkg/sql/opt/optbuilder/window.go, line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgiqyg8CvzdIAyVBQ2q:-Lgiqyg8CvzdIAyVBQ2r:ba53emu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L238)):*
> ```Go
> 	}
> 
> 	b.constructProjectForScope(fromScope, argScope)
> ```

Why do you need to add a project here?
___
*[pkg/sql/opt/optbuilder/window.go, line 308 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgip-OU4kuwUSJN3Bse:-Lgip-OU4kuwUSJN3Bsf:b-z2msdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L308)):*
> ```Go
> // getTypedArgs returns the arguments to the aggregate function as a
> // []tree.TypedExpr similar to getTypedWindowArgs.
> func (b *Builder) getTypedArgs(agg *aggregateInfo) []tree.TypedExpr {
> ```

You could make this function more general by just passing in `agg.Exprs` as a `tree.Exprs` (maybe something like this already exists?)
___
*[pkg/sql/opt/optbuilder/window.go, line 317 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgimED80S-Ikh0mus3X:-LgimED80S-Ikh0mus3Y:b-431cav) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L317)):*
> ```Go
> 
> // Construct the argExprs into a memo.ScalarListExpr.
> func (b *Builder) constructArgs(
> ```

This could use a better name -- maybe `buildWindowArgs`?
___
*[pkg/sql/opt/optbuilder/window.go, line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgimo079UltsJfWhlRg:-Lgimo079UltsJfWhlRh:b-gs7lkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L340)):*
> ```Go
> // given partition and ordering can be added to. If no such frame is found, a
> // new one is made.
> func (b *Builder) findAppropriateFramePosition(
> ```

maybe `findMatchingFrameIndex`?
___
*[pkg/sql/opt/optbuilder/window.go, line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgitTqDA8G0a0XYetYo:-LgitTqDA8G0a0XYetYp:b-wgeapy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L381)):*
> ```Go
> 	windowGroupingColSet := groupingColSet.Copy()
> 
> 	if groupingColSet.Empty() {
> ```

seems like everything inside this `if` statement could be its own helper function
___
*[pkg/sql/opt/optbuilder/window.go, line 403 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgisBeY2YofasbbV24b:-LgisBeY2YofasbbV24c:b-yv8ryz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L403)):*
> ```Go
> 		passthrough := input.Relational().OutputCols
> 		// Deduplicate the columns; we don't need to produce the same aggregation
> 		// multiple times.
> ```

Why would that happen in the first place?
___
*[pkg/sql/opt/optbuilder/window.go, line 414 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgism6B-lVhwgckAKOg:-Lgism6C-KC_QdonOpJt:b-70z6fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L414)):*
> ```Go
> 				aggregateCol := aggInfos[i].col
> 				if requiresProjection {
> 					aggregateCol = b.synthesizeColumn(outScope, aggInfos[i].col.name.String(), aggInfos[i].col.typ, aggInfos[i].col.expr, varExpr)
> ```

can use `aggregateCol` instead of `aggInfos[i].col` everywhere


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> Reviewed 1 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/lease.go, line 474 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgif_A_2ah79J-KRlT0:-Lgz-X_31FLzsuwL_br6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/lease.go#L474)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

stray extra line
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate-opt, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgijRdDANXBd_cfG04S:-Lgz9ObQ-KfPc3c1TmiT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/logictest/testdata/logic_test/aggregate-opt#L1)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

nice tests! Maybe add a couple with concat_agg and string_agg too
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgijcuPAkyzuw3NgTr6:-Lgz9hL_6YldkeR10gvO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/logictest/testdata/logic_test/orms#L27)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You can take out this comment now. Should you also add #25148 to the list of issues that this PR fixes?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/scope.go, line 506 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgikswTDdeg56lmk2_C:-Lgz9oD_9Wc-cmc_sFYf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/scope.go#L506)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] I think this would be slightly more efficient:
```
if match && !agg.IsCommutative() {
  if len(a.OrderBy) != len(agg.OrderBy) {
    match = false
  } else {
    for k := range a.OrderBy {
    ...
```


</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/scope.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgildIACz9HtGf2nguD:-Lgz9otl36Ml6YSCS1Yp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/scope.go#L507)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[super nit] I'd use `j` here instead of `k` since you're only one level deep, not two 
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 696 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgiloZHAmlhskfceRk3:-Lgyr9qF0AXXc6M5Y5J9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8acfe8ffd0028ce1d81a9b1148f7e9ba2673bf95/pkg/sql/opt/optbuilder/select.go#L696)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

stray deleted line
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 67 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgilxN02Z_VPFo14xcQ:-LgyrAl72RS4IGYK_DVd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L67)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[super nit] put start before end
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgio9_73iGtcmwspMzi:-Lgysd6M80coNOZTINrW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L193)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

The calling function, `buildAggregation`, constructs this same `ColSet` later on in that function. I'd move that construction above the call to `buildAggregationAsWindow` so you can just pass in the `ColSet`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 213 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgiq0zwCb_tzRFoPPGD:-Lgytq_F7sc3ujQGSgoa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L213)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

If you're passing `types.Any`, you should probably just use `resolveType`
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgiqSNaD8eX_tcW-Ba9:-LgyskPq9-wSkdR1WXge:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L227)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This stuff can probably be pulled out into a helper function that's called both here and above in `buildWindow`
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 238 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgiqyg8CvzdIAyVBQ2q:-LgytvwGD3YBUjTP0myu:b-xsrrbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L238)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why do you need to add a project here?
</blockquote></details>

(This was in the wrong place) We need one above to construct the pre-projections. Like if the group by is on something like `upper(col)`.
___
*[pkg/sql/opt/optbuilder/window.go, line 308 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgip-OU4kuwUSJN3Bse:-LgywFXA8XsPrSzITvhJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L308)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You could make this function more general by just passing in `agg.Exprs` as a `tree.Exprs` (maybe something like this already exists?)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 317 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgimED80S-Ikh0mus3X:-LgyrmwW2ZiL41Mfv747:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L317)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This could use a better name -- maybe `buildWindowArgs`?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgimo079UltsJfWhlRg:-Lgys00b2pD1faQi-qI7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L340)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

maybe `findMatchingFrameIndex`?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgitTqDA8G0a0XYetYo:-LgzIltE4sorqgnmmzkf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L381)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

seems like everything inside this `if` statement could be its own helper function
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 403 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgisBeY2YofasbbV24b:-LgzJY3l59zlJpLgw2Ne:b-xcvg5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L403)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why would that happen in the first place?
</blockquote></details>

You're right, it wouldn't. Removed.
___
*[pkg/sql/opt/optbuilder/window.go, line 414 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgism6B-lVhwgckAKOg:-Lgz-4zL4XNYuoXSqM-Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L414)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

can use `aggregateCol` instead of `aggInfos[i].col` everywhere
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
This is great, excited to check off another compatibility gap.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-:-Lh06D0n39SusJWkRLur:b-c3nd97)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/opt/optbuilder/groupby.go, line 254 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lh04ap36LUZQte3ohj6:-Lh04ap36LUZQte3ohj7:bf2gog2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c492455d52f8e55344fdaed8518e16297f239984/pkg/sql/opt/optbuilder/groupby.go#L254)):*
> ```Go
> 	}
> 
> 	// If there are any aggregates that are ordering sensitive, build the aggregations
> ```

I expect the window version to be slower. In particular we generate (or at least emit) an aggregated value for each row instead of once per group. We should only do this when we really need to. If we have at most one `ORDER BY` (or the same one shows up multiple times), we should continue to build as a simple group by (which supports a single ordering).
___
*[pkg/sql/opt/optbuilder/window.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lh056oD3h-XAaDC0jN1:-Lh056oD3h-XAaDC0jN2:b-ydxr8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c492455d52f8e55344fdaed8518e16297f239984/pkg/sql/opt/optbuilder/window.go#L207)):*
> ```Go
> // buildAggregationAsWindow builds the aggregation operators as window functions.
> // Returns the output scope for the aggregation operation.
> func (b *Builder) buildAggregationAsWindow(
> ```

It would be useful to show an example with the corresponding tree. You could copy a tree from a testfile (and trim it down to what's relevant)
___
*[pkg/sql/opt/optbuilder/window.go, line 214 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lh04UEGANFpVISRgYnQ:-Lh04UEGANFpVISRgYnR:bw2slyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c492455d52f8e55344fdaed8518e16297f239984/pkg/sql/opt/optbuilder/window.go#L214)):*
> ```Go
> 	aggInfos := aggOutScope.groupby.aggs
> 
> <<<<<<< Updated upstream
> ```

this needs merging


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 13 files at r1, 10 of 12 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @ridwanmsharif)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate-opt, line 73 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhH7hnwEeKsUQHT11hW:-LhH7hnwEeKsUQHT11hX:b-wqejg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/logictest/testdata/logic_test/aggregate-opt#L73)):*
> ```
> 6  aaabbb
> 
> # Testing pre-projections. Tests when the GroupBy clause has a projection.
> ```

[nit] these four tests seem very similar. if they're not testing different things, I'd combine them into one or two tests
___
*[pkg/sql/opt/optbuilder/groupby.go, line 257 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhH8_TPBnPwuB4MLBVy:-LhH8_TQ3VcL5fxlCEzR:bmjtgcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/groupby.go#L257)):*
> ```Go
> 	// as window functions over each group.
> 	if aggOutScope.groupby.HasNonCommutativeAggregates() {
> 		//log.Fatalf(context.Background(), "%+v", aggOutScope.groupby.aggs[0].OrderBy[0])
> ```

stray debug line
___
*[pkg/sql/opt/optbuilder/window.go, line 308 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lgip-OU4kuwUSJN3Bse:-LhHFIiNB-aFUJWA62Op:bqbfgy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ba82afc74f9733cce0cf9520d2940fa96300b6/pkg/sql/opt/optbuilder/window.go#L308)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Now I think this function belongs in `optbuilder/util.go`
___
*[pkg/sql/opt/optbuilder/window.go, line 107 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHAksiEIFrhAmjQ0t1:-LhHAksiEIFrhAmjQ0t2:b-fp40xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L107)):*
> ```Go
> 		}
> 
> 		if w.Filter != nil {
> ```

Do you need to handle FILTER in `buildAggregationAsWindow` too?
___
*[pkg/sql/opt/optbuilder/window.go, line 330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHKGTQ2OFB3ETPUdMr:-LhHKGTQ2OFB3ETPUdMs:becksxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L330)):*
> ```Go
> }
> 
> // buildWindowArgs constructs the argExprs into a memo.ScalarListExpr.
> ```

[nit] you're using the words "build" and "construct" interchangeably here and below. In the past we have tried to make the distinction within `optbuilder` that "build" is used for functions that have relatively complex logic with semantic analysis, etc. (in addition to construction of expressions), while "construct" is reserved for functions that only perform relatively simple, mechanical construction of expressions. It's a subtle difference, but try to be consistent if you can.
___
*[pkg/sql/opt/optbuilder/window.go, line 351 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHFdjn7VJVj1oy4e60:-LhHFdjo4HkAJ6x5kim8:b-cddq84) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L351)):*
> ```Go
> }
> 
> // buildWindowPartition construct the appropriate partitions for window functions.
> ```

[nit] construct -> constructs
___
*[pkg/sql/opt/optbuilder/window.go, line 379 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHG6O6Cv3IE37FxGuf:-LhHG6O6Cv3IE37FxGug:b-cddq84) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L379)):*
> ```Go
> }
> 
> // buildWindowOrdering construct the appropriate orderings for window functions.
> ```

[nit] construct -> constructs
___
*[pkg/sql/opt/optbuilder/window.go, line 448 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHHtLQ57JMS_eDFFkn:-LhHHtLR258ArU5jYTho:b-9ilvfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L448)):*
> ```Go
> 	input memo.RelExpr, groupingColSet opt.ColSet, aggInfos []aggregateInfo, outScope *scope,
> ) memo.RelExpr {
> 	windowGroupingColSet := groupingColSet.Copy()
> ```

[nit] can move this below the if statement below
___
*[pkg/sql/opt/optbuilder/window.go, line 456 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHISGD4T7XcNuLmz3K:-LhHISGD4T7XcNuLmz3L:b-nrnt49) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L456)):*
> ```Go
> 
> 	// Construct a DistinctOn using all the columns added by the window functions
> 	// for the aggregates in addition to the grouping columns.
> ```

Why do you need to add the aggregate columns to the `DistinctOn`? Shouldn't the grouping columns be enough?
___
*[pkg/sql/opt/optbuilder/window.go, line 466 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHJyPWBKh5uTDIJl3n:-LhHJyPWBKh5uTDIJl3o:bvk5sp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L466)):*
> ```Go
> 
> // replaceDefaultReturn constructs a case expression to apply as a projection over
> // a ScalarGroupBy expression, that replaces the default NULL value from natchVal
> ```

natchVal -> matchVal
___
*[pkg/sql/opt/optbuilder/window.go, line 502 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHIf_ADurk5ExHYn8l:-LhHIf_ADurk5ExHYn8m:b-otqj80) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L502)):*
> ```Go
> 	input memo.RelExpr, groupingColSet opt.ColSet, aggInfos []aggregateInfo, outScope *scope,
> ) memo.RelExpr {
> 	windowGroupingColSet := groupingColSet.Copy()
> ```

Why do you need to make a copy? You're not making any changes, right?
___
*[pkg/sql/opt/optbuilder/window.go, line 523 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHIz-26iyXX1TVXwGf:-LhHIz-26iyXX1TVXwGg:b-qkunrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L523)):*
> ```Go
> 
> 	// Create an appropriate passthrough for the projection.
> 	passthrough := input.Relational().OutputCols
> ```

do you need to make a copy of `OutputCols`?


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate-opt, line 73 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhH7hnwEeKsUQHT11hW:-LhaGemEAtVwZDzScwXL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/logictest/testdata/logic_test/aggregate-opt#L73)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] these four tests seem very similar. if they're not testing different things, I'd combine them into one or two tests
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgYpn7AYuj9Zi8IbspH-r1-123:-Lha9PI6DWMi8OYAhMPV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e271e3a5b5e1a3587ceddddcfc4bb53e913a31/pkg/sql/opt/optbuilder/groupby.go#L123)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Ideally, we could reuse `AggregateIsOrderingSensitive` here so we don't have multiple places where we list commutative aggregates.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 254 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lh04ap36LUZQte3ohj6:-LhaGfTV-3ZV85RDtfDX:bm0ej0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c492455d52f8e55344fdaed8518e16297f239984/pkg/sql/opt/optbuilder/groupby.go#L254)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I expect the window version to be slower. In particular we generate (or at least emit) an aggregated value for each row instead of once per group. We should only do this when we really need to. If we have at most one `ORDER BY` (or the same one shows up multiple times), we should continue to build as a simple group by (which supports a single ordering).
</blockquote></details>

You're right. My plan was to ship this as is and then add a separate commit/PR to only use the window functions when absolutely necessary (multiple unique orderings). This could be a normalization rule.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 257 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhH8_TPBnPwuB4MLBVy:-Lha9QbACB3Weg0D6nlp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/groupby.go#L257)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

stray debug line
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 107 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHAksiEIFrhAmjQ0t1:-Lha9TUM2a1It9iQXe_x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L107)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Do you need to handle FILTER in `buildAggregationAsWindow` too?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHKGTQ2OFB3ETPUdMr:-Lha9WMB80H3zUnDpxik:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L330)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] you're using the words "build" and "construct" interchangeably here and below. In the past we have tried to make the distinction within `optbuilder` that "build" is used for functions that have relatively complex logic with semantic analysis, etc. (in addition to construction of expressions), while "construct" is reserved for functions that only perform relatively simple, mechanical construction of expressions. It's a subtle difference, but try to be consistent if you can.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 351 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHFdjn7VJVj1oy4e60:-Lha9_FSDNbqww8Cnd97:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L351)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] construct -> constructs
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 379 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHG6O6Cv3IE37FxGuf:-Lha9_sM4W7-KosAh5Zb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L379)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] construct -> constructs
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 448 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHHtLQ57JMS_eDFFkn:-Lha9azaEI3zqAIS2EJ6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L448)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] can move this below the if statement below
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 456 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHISGD4T7XcNuLmz3K:-Lha9cxaB6UNXh01OujF:blhpf65) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L456)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why do you need to add the aggregate columns to the `DistinctOn`? Shouldn't the grouping columns be enough?
</blockquote></details>

Well because the columns must be part of either an aggregate function or be part of the group by columns. But I see why that was unclear, instead, I'm now using the same `ConstAgg` pattern to be more explicit about this.
___
*[pkg/sql/opt/optbuilder/window.go, line 466 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHJyPWBKh5uTDIJl3n:-Lha9vBbDJWEInydPj7O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L466)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

natchVal -> matchVal
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 502 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHIf_ADurk5ExHYn8l:-LhaGc8gBpE0vLiXvSI2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L502)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why do you need to make a copy? You're not making any changes, right?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 523 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHIz-26iyXX1TVXwGf:-LhaGcaf5U38JluXKAaG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L523)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

do you need to make a copy of `OutputCols`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
Great job! This ended up being a pretty deep rabbit hole but it feels like you figured out all the rough spots! Mostly have some nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-:-LhaMwvSEDIAWb4OkT5k:bpxilkj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, @ridwanmsharif, and @rytaft)

___
*[docs/generated/sql/bnf/stmt_block.bnf, line 1983 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaHNrkA_vAZfgB1kjI:-LhaHNrkA_vAZfgB1kjJ:bxqfhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/feef50b57cfb60ff19f36eaa0aa515eabba030e1/docs/generated/sql/bnf/stmt_block.bnf#L1983)):*
> ```
> 	| func_name '(' expr_list opt_sort_clause ')'
> 	| func_name '(' 'ALL' expr_list opt_sort_clause ')'
> 	| func_name '(' 'DISTINCT' expr_list opt_sort_clause ')'
> ```

what ended up happening with our lack of support for `DISTINCT` in window functions? do we need to exclude this case?
___
*[pkg/sql/logictest/testdata/logic_test/aggregate-opt, line 93 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaHp4T63NJ2ZmF8mqN:-LhaHp4T63NJ2ZmF8mqO:b-br7j19) ([raw file](https://github.com/cockroachdb/cockroach/blob/feef50b57cfb60ff19f36eaa0aa515eabba030e1/pkg/sql/logictest/testdata/logic_test/aggregate-opt#L93)):*
> ```
> # Testing pre-projections. Tests when the GroupBy clause has a projection.
> query IIIT
> SELECT * FROM (SELECT count(1) count_1, count(lower(col3)) count_lower, count(upper(col3)) count_upper, concat_agg(col3 ORDER BY col1) concat from tab GROUP BY upper(col3)) ORDER BY concat
> ```

nit: might be worth sqlfmting this, it's a little hard to follow as-is
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 26 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaHvg4AZjhMMRhhNUu:-LhaHvg4AZjhMMRhhNUv:b-c3ssr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5256fedbc4dd97e2c65989fe8173e777d17045f/pkg/sql/logictest/testdata/logic_test/orms#L26)):*
> ```
> 
> # #12115
> # Skipped until #25148 is solved
> ```

love when these get deleted :thumbsup:
___
*[pkg/sql/opt/optbuilder/groupby.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LgYpn7AYuj9Zi8IbspH-r1-123:-LhaIgPKEQ6Ljdl5B7zJ:bpkv410) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e271e3a5b5e1a3587ceddddcfc4bb53e913a31/pkg/sql/opt/optbuilder/groupby.go#L123)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: could be `return a.OrderBy == nil || !a.IsOrderingSensitive()`
nit: can this be private?
___
*[pkg/sql/opt/optbuilder/groupby.go, line 254 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lh04ap36LUZQte3ohj6:-LhaI7OX63w2jCEblcdy:b-2wlzwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c492455d52f8e55344fdaed8518e16297f239984/pkg/sql/opt/optbuilder/groupby.go#L254)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

You're right. My plan was to ship this as is and then add a separate commit/PR to only use the window functions when absolutely necessary (multiple unique orderings). This could be a normalization rule.
</blockquote></details>

I think this is the philosophically "right" way to do it, I suspect recognizing these cases might end up being a little clunky, but maybe I'm wrong.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 82 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaIxk91JT3fKq0Rf7C:-LhaIxk91JT3fKq0Rf7D:b-e0sg9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/groupby.go#L82)):*
> ```Go
> // HasNonCommutativeAggregates checks whether any of the aggregates are
> // non-commutative or ordering sensitive.
> func (g groupby) HasNonCommutativeAggregates() bool {
> ```

nit: does this need to be exported?
___
*[pkg/sql/opt/optbuilder/groupby.go, line 125 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaIzV1688Af-AfbfTH:-LhaIzV1688Af-AfbfTI:b-k5rcbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/groupby.go#L125)):*
> ```Go
> // ordering sensitive. That is, it can give different results based on the order
> // values are fed to it.
> func (a aggregateInfo) IsOrderingSensitive() bool {
> ```

nit: ditto, does this need to be exported?
___
*[pkg/sql/opt/optbuilder/groupby.go, line 127 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaIpDI3-9TqdF2zDL0:-LhaIpDI3-9TqdF2zDL1:birvd6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/groupby.go#L127)):*
> ```Go
> func (a aggregateInfo) IsOrderingSensitive() bool {
> 	switch a.def.Name {
> 	case "array_agg", "concat_agg", "string_agg", "json_agg", "jsonb_agg":
> ```

don't we already have a function somewhere for this?
___
*[pkg/sql/opt/optbuilder/groupby.go, line 621 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaJBOd29HRjmIQ-v6e:-LhaJBOe4f11tR07kNq3:b-iyrgvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/groupby.go#L621)):*
> ```Go
> 	case "string_agg":
> 		// TODO(ridwanmsharif): Confirm with Justin what the desired behavior here should be.
> 		// Why can we handle non constant second arguments in window fns but not aggregates?
> ```

no good reason, window function execution implements them but agg execution doesn't.
___
*[pkg/sql/opt/optbuilder/window.go, line 97 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaJxiA2UAHBNw7ELaO:-LhaJxiB6HPCFhf3Si18:b-afiopv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L97)):*
> ```Go
> 		// Build appropriate partitions.
> 		windowPartition := b.constructWindowPartition(def.Partitions, i, w.def.Name, inScope, argScope)
> 		partitions[i].UnionWith(windowPartition)
> ```

can't this just be assignment? what's being unioned here?
___
*[pkg/sql/opt/optbuilder/window.go, line 226 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaKc7xA92b7z8mKvjc:-LhaKc7xA92b7z8mKvjd:b-rff0fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L226)):*
> ```Go
> 	partitions := make([]opt.ColSet, len(aggInfos))
> 	orderings := make([]physical.OrderingChoice, len(aggInfos))
> 	windowFrames := make([]tree.WindowFrame, len(aggInfos))
> ```

I don't conceptually understand why we need `tree.WindowFrame` around here, what do we need it for over `memo.WindowFrame`?
___
*[pkg/sql/opt/optbuilder/window.go, line 270 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaL4-V2E1oGqFdi_2G:-LhaL4-V2E1oGqFdi_2H:bmslf0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L270)):*
> ```Go
> 	frames := make([]memo.WindowExpr, 0, len(aggInfos))
> 	for i, agg := range aggInfos {
> 		// Using this instead of constructAggregate so we can have non-constant second
> ```

I don't think this is especially important since we don't support them *without* orderings.
___
*[pkg/sql/opt/optbuilder/window.go, line 288 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaLPpz5BUf9RsUEwrX:-LhaLPpz5BUf9RsUEwrY:b76gzq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L288)):*
> ```Go
> 					Frame: memo.WindowFrame{
> 						Mode:           windowFrames[i].Mode,
> 						StartBoundType: windowFrames[i].Bounds.StartBound.BoundType,
> ```

I think it would be nice if we could just get this down to one "frame" data structure so we don't have this kind of awkward assignment (i say awkward just because we know it will always be unbounded preceding and following)
___
*[pkg/sql/opt/optbuilder/window.go, line 301 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaLq1W3kgzS62zH_1B:-LhaLq1W3kgzS62zH_1C:bnefsdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L301)):*
> ```Go
> 	}
> 
> 	// Construct a grouping so the values per group are squashed down.
> ```

[optional] I get what's going on here but I've also been knee-deep in this stuff for the last month and a half. I suspect without that context this would be sort of hard to grasp, maybe consider adding a little more exposition on what this is doing?
___
*[pkg/sql/opt/optbuilder/window.go, line 501 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaMST4E91WfnPX8ep_:-LhaMST4E91WfnPX8epa:br60qgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L501)):*
> ```Go
> 
> // replaceDefaultReturn constructs a case expression to apply as a projection over
> // a ScalarGroupBy expression, that replaces the default NULL value from natchVal
> ```

nit: `natchVal`
___
*[pkg/sql/opt/optbuilder/testdata/aggregate, line 3178 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaMpAnEaOTiZhlVgk8:-LhaMpAnEaOTiZhlVgk9:b-5kyysd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/testdata/aggregate#L3178)):*
> ```
>       └── rowid int not null (hidden)
> 
> opt
> ```

nit: these should use `build`


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 4 of 6 files at r6, 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/opt/optbuilder/window.go, line 330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHKGTQ2OFB3ETPUdMr:-LhaSrk69FbAjg2dHieA:b2kphc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L330)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

[nit] I'd actually leave this function and the two below as "build" functions, since the work they are doing is non-trivial.
___
*[pkg/sql/opt/optbuilder/window.go, line 466 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHJyPWBKh5uTDIJl3n:-LhaTlbPCwyxB9oHBRnx:b-8dahjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L466)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still there
___
*[pkg/sql/opt/optbuilder/window.go, line 250 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaRWpE12iuuMtqwGv2:-LhaRWpE12iuuMtqwGv3:b-sgx6c0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718c777d9799e1f7566840e65534b7887fd8b84/pkg/sql/opt/optbuilder/window.go#L250)):*
> ```Go
> 		}
> 
> 		if agg.Filter != nil {
> ```

Do you have any tests for this?
___
*[pkg/sql/opt/optbuilder/window.go, line 252 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaS-qg0qIXl5bim1RX:-LhaS-qg0qIXl5bim1RY:bvt0xcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718c777d9799e1f7566840e65534b7887fd8b84/pkg/sql/opt/optbuilder/window.go#L252)):*
> ```Go
> 		if agg.Filter != nil {
> 			defer b.semaCtx.Properties.Restore(b.semaCtx.Properties)
> 			b.semaCtx.Properties.Require("FILTER", tree.RejectSpecial)
> ```

I'd put these lines into `constructFilterCol`


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[docs/generated/sql/bnf/stmt_block.bnf, line 1983 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaHNrkA_vAZfgB1kjI:-Lhb2973Du05j9JGIp7q:b-jkpb1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/feef50b57cfb60ff19f36eaa0aa515eabba030e1/docs/generated/sql/bnf/stmt_block.bnf#L1983)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

what ended up happening with our lack of support for `DISTINCT` in window functions? do we need to exclude this case?
</blockquote></details>

Yeah we do. How do we document this kind of limitation? Have a release note saying no aggregates using distinct can be used in conjunction with ordered aggregates?
___
*[pkg/sql/logictest/testdata/logic_test/orms, line 26 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaHvg4AZjhMMRhhNUu:-LhaUDwp91F5wYkA3TmW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5256fedbc4dd97e2c65989fe8173e777d17045f/pkg/sql/logictest/testdata/logic_test/orms#L26)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

love when these get deleted :thumbsup:
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 127 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaIpDI3-9TqdF2zDL0:-LhasoV6BK_suOybeBra:b-i65ao7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/groupby.go#L127)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

don't we already have a function somewhere for this?
</blockquote></details>

We had `AggregateIsOrderingSensitive`, but I removed that in favor of this so we don't have to build out the ScalarExpr to know whether the aggregate is sensitive to ordering.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 621 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaJBOd29HRjmIQ-v6e:-LhatRBe6pnO6YZ613yk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/groupby.go#L621)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

no good reason, window function execution implements them but agg execution doesn't.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 466 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhHJyPWBKh5uTDIJl3n:-LhatTzGCGQzky6x_E4R:bfveajk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7800a4ced4fa77a01a3e6849efbcf8f0825785e/pkg/sql/opt/optbuilder/window.go#L466)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Still there
</blockquote></details>

Ugh remnants of the git surgery I was doing to amend the commits.
___
*[pkg/sql/opt/optbuilder/window.go, line 97 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaJxiA2UAHBNw7ELaO:-Lhatvxm8pdavkQphOet:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L97)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

can't this just be assignment? what's being unioned here?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 226 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaKc7xA92b7z8mKvjc:-LhaxSQV4zD1cy_KoNc7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L226)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I don't conceptually understand why we need `tree.WindowFrame` around here, what do we need it for over `memo.WindowFrame`?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 270 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaL4-V2E1oGqFdi_2G:-Lhb3cP406mZQIw7JBYf:b-996ok1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L270)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I don't think this is especially important since we don't support them *without* orderings.
</blockquote></details>

We do need to use `constructWindowFn` instead of `constructAggregate` cause with specified orderings `constructAggregate` panics with the multiple non-constant expressions error. Window functions support it so we're good with using it.
___
*[pkg/sql/opt/optbuilder/window.go, line 288 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaLPpz5BUf9RsUEwrX:-LhaxWFu8uHwKwMlBYH6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L288)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think it would be nice if we could just get this down to one "frame" data structure so we don't have this kind of awkward assignment (i say awkward just because we know it will always be unbounded preceding and following)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 301 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaLq1W3kgzS62zH_1B:-LhayXMx4tTVUt6l8Txr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6052b85fa71c48d0fb49a9273c7e88fa5d28d422/pkg/sql/opt/optbuilder/window.go#L301)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

[optional] I get what's going on here but I've also been knee-deep in this stuff for the last month and a half. I suspect without that context this would be sort of hard to grasp, maybe consider adding a little more exposition on what this is doing?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 250 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaRWpE12iuuMtqwGv2:-LhatxtD5fAl1txoz3j8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718c777d9799e1f7566840e65534b7887fd8b84/pkg/sql/opt/optbuilder/window.go#L250)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Do you have any tests for this?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 252 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-LhaS-qg0qIXl5bim1RX:-LhatyPE1CMZ9Vju2iJE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f718c777d9799e1f7566840e65534b7887fd8b84/pkg/sql/opt/optbuilder/window.go#L252)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I'd put these lines into `constructFilterCol`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work!

> Reviewed 6 of 8 files at r10, 2 of 2 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-:-LhaUgrrAy9WZF00tZOd:b-kbynj8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @ridwanmsharif)

___
*[pkg/sql/opt/optbuilder/util.go, line 33 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lhb89mp3Vpc1v31Zhk_:-Lhb89mp3Vpc1v31Zhka:b-v2kek9) ([raw file](https://github.com/cockroachdb/cockroach/blob/962a0a51dfa27c3eee29a21836fc522d9fd27cb3/pkg/sql/opt/optbuilder/util.go#L33)):*
> ```Go
> }
> 
> // windowAggregateFrame() returns a frame that any aggregate build as a window
> ```

[nit] build -> built
___
*[pkg/sql/opt/optbuilder/window.go, line 288 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lhb8ZyE5mUI7i28wujG:-Lhb8ZyE5mUI7i28wujH:b-hnl61i) ([raw file](https://github.com/cockroachdb/cockroach/blob/962a0a51dfa27c3eee29a21836fc522d9fd27cb3/pkg/sql/opt/optbuilder/window.go#L288)):*
> ```Go
> 	// aggregations built as window functions emit an aggregated value for each row
> 	// instead of each group. To rectify this, we must 'squash' the values down by
> 	// wrapping it around a GroupBy or ScalarGroupBy.
> ```

[nit] wrapping it around -> wrapping it with


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-:-LheokB10JRF8dFMBXhP:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/optbuilder/util.go, line 33 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lhb89mp3Vpc1v31Zhk_:-LheohThDOduEOOneSGV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/962a0a51dfa27c3eee29a21836fc522d9fd27cb3/pkg/sql/opt/optbuilder/util.go#L33)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] build -> built
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 288 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/38014#-Lhb8ZyE5mUI7i28wujG:-LheoifXDkB-ylkdAKPh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/962a0a51dfa27c3eee29a21836fc522d9fd27cb3/pkg/sql/opt/optbuilder/window.go#L288)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] wrapping it around -> wrapping it with
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Great work!
